harness/determined/experimental/inference/_torch_batch_process.py

```diff
@@ -0,0 +1,272 @@
+
+        per_batch_processor = batch_processor_cls(
+            init_info=TorchBatchInitInfo(
+                rank=rank,
+                local_rank=core_context.distributed.get_local_rank(),
+                size=core_context.distributed.get_size(),
+                num_agents=core_context.distributed.get_num_agents(),
+                default_device=get_default_device(core_context),
+                tensorboard_path=core_context.train.get_tensorboard_path(),
+            ),
+        )
+
+        # Check if previous checkpoint exists
+        if latest_checkpoint is not None:
+            logging.info("Checkpoint is not none")
+            with core_context.checkpoint.restore_path(latest_checkpoint) as path:
+                metadata = _load_state(path)
+                skip = metadata["steps_completed"]
+                logging.info(f"Previous run completed {skip} steps")
+
+        dataloader = DataLoader(
+            dataset=dataset, batch_size=batch_size, shuffle=False, **dataloader_kwargs
+        ).get_data_loader(repeat=False, skip=skip, num_replicas=total_worker, rank=rank)
+
+        # Create dummy searcher op to report progress to master
+        dummy_searcher_op = None
+        # Initialize dummy searcher for progress report
+        if rank == 0:
+            dummy_searcher_op = core.DummySearcherOperation(1, True)
+
+        dataloader_iterator = iter(dataloader)
+
+        # Enumerate over dataloader directly may cause some workers to iterate for 1 more time
+        # than others when drop_last = False. If those workers synchronize on the last batch_idx,
+        # they would hang forever as other workers never hit that last batch_idx.
+        # To avoid the issue, we calculate and take the ceiling of the iteration count to ensure
+        # all workers iterate for the same number of times.
+        max_batch = math.ceil(dataset_len / batch_size / total_worker)
+        iterate_length = _validate_iterate_length(iterate_length, max_batch)
+
+        last_checkpoint_idx = -1
+
+        for batch_idx in range(skip, iterate_length):
+            logging.info(f"Currently processing batch {batch_idx}")
+            X = next(dataloader_iterator, None)
+            if X is not None:
+                per_batch_processor.process_batch(batch=X, batch_idx=batch_idx)
+
+            if (batch_idx + 1) % checkpoint_interval == 0:
+                _synchronize_and_checkpoint(core_context, batch_idx, rank)
+                last_checkpoint_idx = batch_idx
+                # Report progress can only be done accurately with synchronization
+                core_context._tensorboard_manager.sync(mangler=get_rank_aware_path)
+                if rank == 0:
+                    _report_progress_to_master(
+                        dummy_searcher_op, batch_idx, total_worker, batch_size, dataset_len
+                    )
+
+            if (batch_idx + 1) % check_preempt_interval == 0:
+                if core_context.preempt.should_preempt():
```

Comment:
in the training world, we allow users to set preemption themselves, with `context.set_stop_requested`. i wonder if this is something we should support for inference as well?

Reply:
- What are the major use cases where user would want to request stopping within code from training
- Do we expect all workers to report stopping at the same time, or maybe only some worker will request stopping?